be/src/vec/data_types/data_type_array.cpp

```diff
@@ -188,77 +189,64 @@ std::string DataTypeArray::to_string(const IColumn& column, size_t row_num) cons
+    if (json_value.type() != simdjson::ondemand::json_type::array) {
+        return Status::InvalidArgument("Parse json data failed, not array type '{}'",
+                                       json_value.type().take_value());
     }
-
-    // ltrim
-    while (!rb.eof() && isspace(*rb.position())) {
-        ++rb.position();
-        element.data = rb.position();
-    }
-
-    // parse string
-    if (*rb.position() == '"' || *rb.position() == '\'') {
-        const char str_sep = *rb.position();
-        size_t str_len = 1;
-        // search until next '"' or '\''
-        while (str_len < rb.count() && *(rb.position() + str_len) != str_sep) {
-            ++str_len;
-        }
-        // invalid string
-        if (str_len >= rb.count()) {
-            rb.position() = rb.end();
-            return false;
-        }
-        has_quota = true;
-        rb.position() += str_len + 1;
-        element.size += str_len + 1;
-    }
-
-    // parse array element until array separator ',' or end ']'
-    while (!rb.eof() && (*rb.position() != ',') && (rb.count() != 1 || *rb.position() != ']')) {
-        // invalid elements such as ["123" 456,"789" 777]
-        // correct elements such as ["123"    ,"789"    ]
-        if (has_quota && !isspace(*rb.position())) {
-            return false;
+    simdjson::ondemand::array outer_array = json_value.get_array();
+    auto* array_column = assert_cast<ColumnArray*>(column);
+    auto& offsets = array_column->get_offsets();
+    IColumn& nested_column = array_column->get_data();
+    DCHECK(nested_column.is_nullable());
+    auto& nested_null_col = reinterpret_cast<ColumnNullable&>(nested_column);
+    bool is_string_nested = is_string(remove_nullable(nested));
+    size_t element_num = 0;
+    for (auto it = outer_array.begin(); it != outer_array.end(); ++it) {
+        Status st;
+        try {
+            if (is_complex_type(remove_nullable(nested))) {
+                simdjson::ondemand::value val;
+                auto error_code = (*it).get(val);
+                if (simdjson::SUCCESS != (*it).get(val)) {
+                    st = Status::InvalidArgument(
+                            "Parse json data failed, error code: {}, error "
+                            "info: {}",
+                            error_code, simdjson::error_message(error_code));
+                } else {
+                    st = nested->from_json(val, &nested_null_col);
+                }
+            } else {
+                std::string_view sv = simdjson::trim((*it).raw_json_token().value());
+                if (is_string_nested) {
+                    StringRef sr(sv.data(), sv.size());
+                    StringRef del("\"");
+                    sv = simd::VStringFunctions::trim(sr, del);
+                }
+                ReadBuffer nested_rb(const_cast<char*>(sv.data()), sv.size());
+                st = nested->from_string(nested_rb, &nested_column);
```

Comment:
i don't think it's quite reasonable to call from_string here, `from_json` semantic is different from `from_string`

Reply:
mix them could be miss leading